feat: implement `toLocaleString` method in `dstructs/named-typed-tuple`

Gauravkaushik-1206 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. 
report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: na
  - task: lint_package_json status: na
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: na
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: na
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed
---

Resolves #6976

## Description

This pull request implements the `toLocaleString()` function in  
`lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js`,  
which was flagged by ESLint under the `no-warning-comments` rule.

## Related Issues

- Resolves #6976

## Questions

I have written test cases for this function. Would you like me to include them in this PR or submit them separately?

## Other Information

No additional information.

## Checklist

- [x] Read, understood, and followed the [contributing guidelines][contributing].
- [x] Static analysis checks have passed.
- [x] Code changes are consistent with project style guidelines.

---

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Gauravkaushik-1206 committed:
feat(named-typed-tuple): implement toLocaleString function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,33 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+			if ( this !== tuple ) { // eslint-disable-line no-invalid-this
+				throw new TypeError( 'invalid invocation. `this` is not host tuple.' );
+			}
+			for ( i = 0; i < nfields; i++ ) {
```
@Gauravkaushik-1206 Thank you for working on this. This is a good start. Mind adding argument validation logic similar to https://github.com/stdlib-js/stdlib/blob/fe96d973306fc6d28936c6150c160fb5d1f590c7/lib/node_modules/%40stdlib/array/complex64/lib/main.js#L2620?

Also, similar to that implementation, we can assume that `toLocaleString` exists and can avoid the nested conditional below.

Gauravkaushik-1206 replied:
Sure, I will add the argument validation logic.

kgryte committed:
Discard changes to lib/node_modules/@stdlib/dstructs/named-typed-tuple/examples/index.js

kgryte commented:
@Gauravkaushik-1206 You can include tests in this PR. Same thing for adding docs, benchmarks, etc.

headlessNode committed:
fix: update `fromIndex` handling in `blas/ext/base/ndarray/slast-index-of`

PR-URL: https://github.com/stdlib-js/stdlib/pull/8007
Closes: https://github.com/stdlib-js/metr-issue-tracker/issues/71
Ref: https://github.com/stdlib-js/stdlib/issues/2656
Reviewed-by: Athan Reines <kgryte@gmail.com>

headlessNode committed:
fix: update `fromIndex` handling in `blas/ext/base/ndarray/glast-index-of`

PR-URL: https://github.com/stdlib-js/stdlib/pull/8008
Closes: https://github.com/stdlib-js/metr-issue-tracker/issues/72
Ref: https://github.com/stdlib-js/stdlib/issues/2656
Reviewed-by: Athan Reines <kgryte@gmail.com>

Gauravkaushik-1206 committed:
feat: add argument validation to toLocaleString

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Gauravkaushik-1206 commented:
docs, benchmarks already include the toLocaleString() and i have included the tests.

Gauravkaushik-1206 committed:
feat: changed copyright year to 2025

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Gauravkaushik-1206 commented:
any update ??

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/README.md

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/benchmark/benchmark.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/docs/types/index.d.ts

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/examples/index.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/lib/index.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/lib/main.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/glast-index-of/test/test.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/README.md

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/benchmark/benchmark.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/docs/types/index.d.ts

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/examples/index.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/lib/index.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/lib/main.js

kgryte committed:
Discard changes to lib/node_modules/@stdlib/blas/ext/base/ndarray/slast-index-of/test/test.js

kgryte commented:
/stdlib merge

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
 
-		// TODO: consider adding `toLocaleString()` in a manner similar to `toString()` below
+		/**
+		* Serializes a tuple as a locale-aware string.
```
```suggestion
		* Serializes a tuple as a locale-specific string.
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+		*
+		* @private
+		* @memberof tuple
+		* @param {(string|string[])} [locales] - locale(s)
```
```suggestion
		* @param {(string|Array<string>)} [locales] - locale identifier(s)
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+		* @private
+		* @memberof tuple
+		* @param {(string|string[])} [locales] - locale(s)
+		* @param {Object} [options] - formatting options
```
```suggestion
		* @param {Object} [options] - configuration options
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+		* @memberof tuple
+		* @param {(string|string[])} [locales] - locale(s)
+		* @param {Object} [options] - formatting options
+		* @throws {TypeError} `this` must be the host tuple
```
```suggestion
		* @throws {TypeError} `this` must be the host tuple
		* @throws {TypeError} first argument must be a string or an array of strings
		* @throws {TypeError} options argument must be an object
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+		* @param {(string|string[])} [locales] - locale(s)
+		* @param {Object} [options] - formatting options
+		* @throws {TypeError} `this` must be the host tuple
+		* @returns {string} tuple locale string representation
```
```suggestion
		* @returns {string} string representation
```

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix-tolocale-string

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+			if ( this !== tuple ) { // eslint-disable-line no-invalid-this
+				throw new TypeError( 'invalid invocation. `this` is not host tuple.' );
+			}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+			var out;
+			var val;
+			var i;
+			if ( this !== tuple ) { // eslint-disable-line no-invalid-this
```
```suggestion

			if ( this !== tuple ) { // eslint-disable-line no-invalid-this
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+			} else {
+				throw new TypeError( format( 'invalid argument. First argument must be a string or an array of strings. Value: `%s`.', locales ) );
+			}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,48 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+				val = tuple[ indices[ i ] ];
+				out.push( val.toLocaleString( loc, opts ) );
+			}
+			return out.join(',');
```
```suggestion
			return out.join( ',' );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/lib/main.js:
```diff
@@ -1198,7 +1199,49 @@ function factory( names, options ) { // eslint-disable-line max-lines-per-functi
+			out = [];
+			for ( i = 0; i < nfields; i++ ) {
+				val = tuple[ indices[ i ] ];
+				out.push( val.toLocaleString( loc, opts ) );
```
@Gauravkaushik-1206 What you've implemented is the equivalent of array serialization; however, if you notice in the `toString` method below, we actually want to serialize tuples a bit differently. Namely, we want to preserve the field names. In short, I suggest copying L1260-1269 and then calling `toLocaleString` on the field names and tuple values. E.g.,

```
out = opts.name.toLocaleString( loc, opts ) + '(';
for ( i = 0; i < nfields; i++ ) {
	out += fields[ i ].toLocaleString( loc, opts );
	out += '=';
	out += tuple[ indices[ i ] ].toLocaleString( loc, opts );
	if ( i < nfields-1 ) {
		out += ', ';
	}
}
out += ')';
```

kgryte replied:
You'll want to update the tests accordingly.

Gauravkaushik-1206 replied:
Thank you for the feedback!
I implemented array-style serialization because both the index.d.ts type definitions and the documentation currently specify that the output should be a comma-separated list of values (array-like format).
Do you want to update them also?

<img width="547" height="439" alt="Screenshot 2025-09-08 at 3 13 40 AM" src="https://github.com/user-attachments/assets/cbc580d7-d9b6-47c2-a978-33a6e25892c4" />

<img width="561" height="533" alt="Screenshot 2025-09-08 at 3 14 29 AM" src="https://github.com/user-attachments/assets/7c74627c-5a30-4290-9e33-087a5a695678" />

kgryte replied:
Yeah, I think so.

Gauravkaushik-1206 replied:
okay

kgryte replied:
Just for context, the reason why the docs show that is simply because we had yet to implement the custom logic.

Gauravkaushik-1206 replied:
That makes sense. I’ll proceed with updating the implementation.

kgryte commented:
@Gauravkaushik-1206 Be careful about manipulating Git history. I had to discard a number of unrelated changes when reviewing this PR. For this PR, you shouldn't need to be merging any changes from the upstream repository. That may change in future PRs, but for this one, I'd suggest not worrying about keeping changes in sync with the latest `develop`. You will however need to pull down the changes I've made to this PR before making any additional changes on your end. We discuss how to do this in our contributing guidelines.

Gauravkaushik-1206 committed:
fix: change to custom serialization instead of array-style serialization

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
/stdlib merge

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/docs/repl.txt:
```diff
@@ -1378,10 +1378,11 @@ tuple.toLocaleString( [locales[, options]] )
     Examples
     --------
-    > var factory = {{alias}}( [ 'x', 'y', 'z' ] );
+    > opts = { 'name': 'Point' };
```
```suggestion
    > var opts = { 'name': 'Point' };
```

kgryte committed:
docs: fix missing declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix missing declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: rename variable

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,204 @@
+	t.end();
+});
+
+tape( 'the method handles null and undefined values using String()', function test( t ) {
```
```suggestion
tape( 'the method handles null and undefined values', function test( t ) {
```

kgryte committed:
test: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,204 @@
+
+	t.strictEqual( actual, expected, 'uses the custom toLocaleString method from the object' );
+	t.end();
+});
```
```suggestion
```

This isn't correct, as the backing data store is a typed array. Meaning, everything must serialize to a number, and the accessed value is always a primitive number.

kgryte committed:
test: remove test

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,183 @@
+	t.strictEqual( actual, expected, 'calls toLocaleString on the Date object' );
+	t.end();
+});
+
```
```suggestion
```

These tests are not applicable. The backing data store is a numeric typed array.

kgryte committed:
test: remove tests

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,149 @@
+	t.end();
+});
+
+tape( 'the method handles boolean and NaN values', function test( t ) {
```
```suggestion
tape( 'the method handles NaN values', function test( t ) {
```

kgryte committed:
test: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,149 @@
+	var r;
+
+	Record = namedtypetuple( [ 'isValid', 'value' ] );
+	r = new Record( [ true, NaN ] );
```
```suggestion
	r = new Record( [ 1, NaN ] );
```

kgryte committed:
test: update test value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: rename variable and update test messages

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: reorder tests

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/dstructs/named-typed-tuple/test/test.to_locale_string.js:
```diff
@@ -0,0 +1,148 @@
+		[]
+	];
+	for ( i = 0; i < values.length; i++ ) {
+		t.throws( badValue( values[i] ), TypeError, 'throws a TypeError when provided '+values[i] );
```
```suggestion
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
```

kgryte committed:
test: update message

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix-tolocale-string
